### ISSUE TITLE
intersperseM offsets every interspersed action by 1

### ISSUE THREAD

**Issue by Fybe** [composewell/streamly#1039]
The output of
```haskell
import Data.Function ((&))
import Streamly.Prelude as S

main :: IO ()
main
  = S.fromList [1,2,3,4]
  & S.mapM print
  & S.intersperseM (putStrLn "---")
  & S.drain
```
is
```
1
2
---
3
---
4
---
```
which is obviously not what you would expect.

Tested with `streamly-0.7.2` on `lts-16.12` and `streamly-0.7.3` on `lts-17.8`.

**Comment by adithyaov** [comment]
This might be a limitation w.r.t streaming.

Given how the streams are designed in streamly, the action needs
to occur while trying to get the next element. The only way we
know an element exists is by running the action.

Since intersperse needs to be aware of stream termination the
interspersed action is run only after checking for termination
hence every interspersed action is offset by one.

**Comment by adithyaov** [comment]
We should probably document this behaviour as a warning so users
aren't caught offguard.

**Comment by Fybe** [comment]
Ahh that makes sense! An alternative might be to run the action after every element encountered, so my example would look like this
```
1
---
2
---
3
---
4
---
```
But this behavior can easily be replicated with mapM. So documenting this behavior is probably the better choice.

**Comment by adithyaov** [comment]
`interspeseM` is true to it's name w.r.t the generated values
but (unfortunately) not w.r.t the order of the effects.

Running an action after every element has an undesirable
behaviour of producing an extra effect. Effects in general can be
pretty costly (For eg, a server fetch).

**Comment by harendra-kumar** [comment]
Yes, it intersperses the results but not the effects. We should document it properly.

**Comment by harendra-kumar** [comment]
I guess this is fixed by #1490 .
